File: languages/go/entity/key/managed_key.go
package key

import "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"

// ManagedKey represents a managed key.
type ManagedKey struct {
	// ID is the unique identifier of the managed key (ex: 46c49ee7-ef44-472c-a873-ce81a2d5d764).
	ID string
	// Name is the name of the managed key.
	Name string
	// Protection is the protection level for the key.
	Protection KeyProtectionLevel
	// KeyType is the type of the key.
	KeyType KeyType
	// Expiration is the timestamp indicating when the key expires.
	Expiration int64
	// Key is the actual public key.
	Key string
	// AccessControlType is the access control type for the key.
	AccessControlType AccessControlType
}

func NewManagedKeyFromProto(s *proto.ManagedKey) ManagedKey {
	if s == nil {
		return ManagedKey{}
	}
	return ManagedKey{
		ID:                s.Id,
		Name:              s.GetName(),
		Protection:        KeyProtectionLevelFromProto[s.Protection],
		KeyType:           KeyTypeFromProto[s.KeyType],
		Expiration:        s.GetExpiration(),
		Key:               s.Key,
		AccessControlType: AccessControlTypeFromProto[s.AccessControlType],
	}
}

func (s ManagedKey) ToProto() *proto.ManagedKey {
	return &proto.ManagedKey{
		Id:                s.ID,
		Name:              s.Name,
		Protection:        KeyProtectionLevelToProto[s.Protection],
		KeyType:           KeyTypeToProto[s.KeyType],
		Expiration:        s.Expiration,
		Key:               s.Key,
		AccessControlType: AccessControlTypeToProto[s.AccessControlType],
	}
}

func (s ManagedKey) Type() KeyType {
	return s.KeyType
}

File: languages/js/src/entity/key/managed_key.ts
import * as keysEntitiesProto from "../../bridge/proto/bloock_keys_entities";
import { AccessControlType } from "./access_control_type";
import { KeyProtectionLevel } from "./key_protection_level";
import { KeyType } from "./key_type";

/**
 * Represents a managed key.
 */
export class ManagedKey {
  /**
   * Is the unique identifier of the managed key (ex: 46c49ee7-ef44-472c-a873-ce81a2d5d764).
   */
  public id: string;
  /**
   * Is the name of the managed key.
   */
  public name?: string;
  /**
   * Is the protection level for the key.
   */
  public protection: KeyProtectionLevel;
  /**
   * Is the type of the key.
   */
  public keyType: KeyType;
  /**
   * Is the timestamp indicating when the key expires.
   */
  public expiration?: number;
  /**
   * Is the actual public key.
   */
  public key: string;
  /**
   * Is the access control type for the key.
   */
  public accessControlType: AccessControlType;

  /**
   * Constructs a ManagedKey object with the specified parameters.
   * @param id
   * @param protection
   * @param keyType
   * @param key
   * @param accessControlType
   * @param name
   * @param expiration
   */
  constructor(
    id: string,
    protection: KeyProtectionLevel,
    keyType: KeyType,
    key: string,
    accessControlType: AccessControlType,
    name?: string,
    expiration?: number
  ) {
    this.id = id;
    this.protection = protection;
    this.keyType = keyType;
    this.key = key;
    this.accessControlType = accessControlType;
    this.name = name;
    this.expiration = expiration;
  }

  public toProto(): keysEntitiesProto.ManagedKey {
    return keysEntitiesProto.ManagedKey.fromPartial({
      id: this.id,
      protection: KeyProtectionLevel.toProto(this.protection),
      keyType: KeyType.toProto(this.keyType),
      key: this.key,
      name: this.name,
      expiration: this.expiration,
      accessControlType: AccessControlType.toProto(this.accessControlType)
    });
  }

  static fromProto(r: keysEntitiesProto.ManagedKey): ManagedKey {
    return new ManagedKey(
      r.id,
      KeyProtectionLevel.fromProto(r.protection),
      KeyType.fromProto(r.keyType),
      r.key,
      AccessControlType.fromProto(r.accessControlType),
      r.name,
      r.expiration
    );
  }
}
